src/modm/driver/encoder/encoder.hpp

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) 2021, Thomas Sommer
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#ifndef MODM_ENCODER_HPP
+#define MODM_ENCODER_HPP
+
+#include <modm/platform.hpp>
+#include <type_traits>
+
+namespace modm
+{
+/**
+ * \brief	Encoder
+ *
+ * \ingroup	modm_ui_encoder
+ * \author	2021 Thomas Sommer
+ */
+
+/// @ingroup	modm_driver_encoder
+struct encoder
+{
+	typedef enum Division
+	{
+		ONE,
+		TWO,
+		FOUR
+	} Division_t;
+};
+
+template<typename PH_A, typename PH_B, encoder::Division_t DIVISION,
+		 typename UnderlyingType = int8_t>
+class Encoder
+{
+	static_assert(std::is_signed<UnderlyingType>::value and std::is_integral<UnderlyingType>::value,
+				  "UnderlyingType must be signed integral");
+
+	using Phases = modm::platform::SoftwareGpioPort<PH_A, PH_B>;
+
+public:
+	Encoder() : enc_delta(0), raw_last(0){};
+
+	inline void
+	connect()
+	{
+		Phases::setInput(::Gpio::InputType::PullUp);
+
+		// Tare power-on state
+		modm::delay(10us);
+		UnderlyingType raw = 0;
+		auto read = Phases::read();
```

Comment:
This variable could be const
```suggestion
		const auto read = Phases::read();
```